### PR TITLE
AIs can now select portraits as their display

### PR DESCRIPTION
**Pull request by tralezab**
## About The Pull Request

A new option has been added to the AI display radial menu alongside random, "Portrait"

It opens a small menu that lets you peruse all of the portraits and select one as your display.

image

## Why It's Good For The Game
This can let avid artists draw exactly what they want their AI to be like, and not-so-avid artists to bring tears to my eyes.

Pictures soon!

## Changelog
:cl:
add: AIs can now select art from the gallery as their display!
/:cl:

### CODE REVIEW

**Comment by tralezab** [comment]
Pictures will be added, if you can forgive my painting skills.

File: code/modules/asset_cache/asset_list_items.dm [hunk 1]
```diff
@@ -466,3 +466,27 @@
 /datum/asset/spritesheet/mafia/register()
 	InsertAll("", 'icons/obj/mafia.dmi')
 	..()
+
+/datum/asset/simple/portraits
+	var/tab = "stop"
+	assets = list(
+	)
```
Comment (Jared-Fogle) [inline code/modules/asset_cache/asset_list_items.dm:473]:
```suggestion
	assets = list()
```

File: code/modules/asset_cache/asset_list_items.dm [hunk 2]
```diff
@@ -466,3 +466,27 @@
...
+
+/datum/asset/simple/portraits/New()
+	if(!SSpersistence.paintings || !SSpersistence.paintings[tab] || !length(SSpersistence.paintings[tab]))
+		return
+	for(var/p in SSpersistence.paintings[tab])
+		var/list/portrait = p
+		var/png = "data/paintings/[tab]/[portrait["md5"]].png"
+		if(fexists(png))
+			assets["[portrait["title"]]"] = png
```
Comment (Jared-Fogle) [inline code/modules/asset_cache/asset_list_items.dm:482]:
```suggestion
			assets[portrait["title"]] = png
```

**Comment by necromanceranne** [comment]
The best ai portrait.

Catgirl butt.

**Comment by ATH1909** [comment]
InB4 people make an AI portrait that looks like an empty/inactive AI core.

**Comment by tralezab** [comment]
They don't have animations so don't worry

**Comment by tralezab** [comment]
well i mean at least that stops SOME of the clownery

**Comment by CRITAWAKETS** [comment]
i now have an excuse to draw mr. progs everywhere

infact i already had an excuse and i should work on that now

**Comment by cacogen** [comment]
Can you post a screenshot of it on the mob?

**Comment by tralezab** [comment]
image
ai that kind of wishes it wasn't there

**Comment by optimumtact** [comment]
why did this PR ping me reee

**Comment by tralezab** [comment]
wanna like my pr?

**Comment by tralezab** [comment]
Okay so the solution to the 23x23 on a 32x32 issue never being able to be centered is to introduce a new canvas called 24x25 that perfectly fits the AI. during testing, 23x23 will be available but for final merge i'm gonna swap it to 24x25

**Comment by Timberpoes** [comment]
image

**Comment by ATH1909** [comment]
Can you choose your death display as well?

If so, is it possible to change your display while dead?

**Comment by tralezab** [comment]
death display is now the default

**Comment by Winterous** [comment]
I love this and I want some sort of art browser console to see all the stuff that can show up in galleries as well!

File: tgui/packages/tgui/interfaces/PortraitPicker.js [hunk 3]
```diff
@@ -120,10 +122,10 @@ export const PortraitPicker = (props, context) => {
             <Flex.Item mt={1}>
               <NoticeBox info>
                 {multiline`
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/PortraitPicker.js:124]: Doesn't need the multiline string, you can just shove the text directly into the tag.

**Comment by Cyprex** [comment]
"lewd" is OOC, should be something more official-sounding imo

**Comment by Jared-Fogle** [comment]
Seems like a nonissue, but I think "provocative" fits better anyway.

**Comment by wabash1** [comment]
This is cool but it needs the ability to change status display to the art as well. the only way people can see your picture is admeme or that traitor that has come to card you.

File: code/modules/mob/living/silicon/ai/ai_portrait_picker.dm [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+
+#define TAB_LIBRARY 1
+#define TAB_SECURE 2
+#define TAB_PRIVATE 3
+
+/datum/portrait_picker
+	var/client/holder //client of whoever is using this datum
+
+/datum/portrait_picker/New(user)//user can either be a client or a mob due to byondcode(tm)
```
Comment (AnturK) [inline code/modules/mob/living/silicon/ai/ai_portrait_picker.dm:15]: You're doing something wrong if that's the case.

File: code/modules/mob/living/silicon/ai/ai_portrait_picker.dm [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+				MA.pixel_x = 5
+				MA.pixel_y = 5
+			else if(w == 24 || h == 24)
+				to_chat(ai, "<span class='notice'>Portrait Accepted. Enjoy!</span>")
+				MA.pixel_x = 4
+				MA.pixel_y = 4
+			else
+				to_chat(ai, "<span class='warning'>Sorry, only 23x23 and 24x24 Portraits are accepted.</span>")
+				return
+			ai.cut_overlays() //so people can't keep repeatedly select portraits to add stacking overlays
```
Comment (AnturK) [inline code/modules/mob/living/silicon/ai/ai_portrait_picker.dm:76]: Use update overlays to pass back single one if it's set instead ? This applies to most cut_overlays() added in this pr.

File: code/modules/asset_cache/asset_list_items.dm [hunk 6]
```diff
@@ -466,3 +466,26 @@
...
 	..()
+
+/datum/asset/simple/portraits
+	var/tab = "use subtypes of this please"
+	assets = list()
+
+/datum/asset/simple/portraits/New()
+	if(!SSpersistence.paintings || !SSpersistence.paintings[tab] || !length(SSpersistence.paintings[tab]))
+		return
+	for(var/p in SSpersistence.paintings[tab])
```
Comment (AnturK) [inline code/modules/asset_cache/asset_list_items.dm:477]: I'm not sure loading and sending all portraits at once is a good idea.

**Review by Timberpoes — CHANGES_REQUESTED** [review]
image

Make sure you run the tgui linter before your final commit and fix everything that pops up. Ideally you should run it before you push any commits, too.

**Review by Twaticus — APPROVED** [review]

**Comment by wabash1** [comment]
Is the private tab working? It crashed tgui when I last used it.

**Comment by tralezab** [comment]
@wabash1 yeah, just that now it doesn't show the tab if there's nothing in it.

File: tgui/packages/tgui/interfaces/PortraitPicker.js [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    <Window
+      theme="ntos"
+      title="Portrait Picker"
+      width={400}
+      height={406}>
+      <Window.Content>
+        <Flex height="100%" direction="column">
+          <Flex.Item mb={1}>
+            <Section fitted>
+              <Tabs fluid={1}>
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/PortraitPicker.js:42]:
```suggestion
              <Tabs fluid textAlign="center">
```

**Comment by tralezab** [comment]
i hate how basically ready this is

**Comment by stylemistake** [comment]
sigh, do i really need to rebuild this myself?

**Comment by tralezab** [comment]
no i'll do it for you, but we can pretty much merge this the moment i do. people need to make more 24x24 paintings in game

**Comment by Winterous** [comment]
People would probably make more art if there was better reward for doing so; it's just a rare chance of seeing it in later rounds, no other way of seeing it ever again.

**Comment by wabash1** [comment]
@winterous I would think to encourage more art there needs to be a search
function in this for portraits, and then later someone could add the
ability for curators to print paintings for a price and that same search
function would be used again. Also a rule against spamming duplicates would
help and a general gallery people can view on web.

**Comment by stylemistake** [comment]
rip, i should've rebuilt it myself

**Pull request merged**